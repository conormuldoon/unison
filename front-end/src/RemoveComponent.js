
import PropTypes from 'prop-types';
import React from 'react';
import { problemConnecting } from './Util';
import HttpStatus from 'http-status-codes';


export function createRemoveFactory(obtainData, href, name) {

  return function removeFactory(hideDisplay) {
      return <RemoveComponent obtainData={obtainData}
          hideDisplay={hideDisplay}
          href={href}
          name={name}
      />;
  }
}

/**
 * A component to enable the user to remove a location from being tracked.
 * 
 * @component
 */
function RemoveComponent({obtainData, hideDisplay, href, name }) {

  async function removeRequest() {

    const response = await fetch(href, {
      method: 'DELETE'
    });

    if (response.ok) {

      alert(name + ' was removed.');
      obtainData();

    } else if (response.status === HttpStatus.UNAUTHORIZED) {
      alert('Incorrect user name or password');
    } else if (response.status === HttpStatus.FORBIDDEN) {
      alert('You do not have permission to delete ' + name + '.');
    } else {
      console.log(response.status)
      problemConnecting();
    }

  }

  function handleClick() {
    hideDisplay();
    if (window.confirm("Are you sure you would like to remove " + name +
      "? This will also delete all of the harvested data assoicated with " + name + ".")) {
      removeRequest();
    }
  }


  return (

    <button data-testid='rm-button' onClick={handleClick}>{'Remove ' + name}</button>

  );

}


RemoveComponent.propTypes = {

  /** Called when the remove button is clicked to hide the add location form if it is displayed. */
  hideDisplay: PropTypes.func.isRequired,

  /** Called when a location has been successfully removed to obtain the updated location list from the server. */
  obtainData: PropTypes.func.isRequired,

  name: PropTypes.string.isRequired,

  href: PropTypes.string.isRequired,
}

export default RemoveComponent;
