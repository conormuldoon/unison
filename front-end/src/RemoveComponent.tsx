

import React from 'react';
import { problemConnecting, csrfToken } from './Util';
import HttpStatus from 'http-status-codes';
import PropTypes from 'prop-types';


export function createRemoveFactory(obtainData: () => void, name: string, href?: string) {

  return function removeFactory(hideAdd: () => void): JSX.Element | undefined {
    if (href) {
      return <RemoveComponent obtainData={obtainData}
        hideAdd={hideAdd}
        href={href}
        name={name}
      />;
    }
  }
}

export interface RemoveProps {

  /** Called when a location has been successfully removed to obtain the updated location list from the server. */
  obtainData: () => void;

  /** Called when the remove button is clicked to hide the add location form if it is displayed. */
  hideAdd: () => void;


  /** The URI used to delete the lcoation. */
  href: string;

  /** The name of the currently selected location. */
  name: string;

}
/**
 * A component to enable the user to remove a location from being tracked.
 * 
 * @component
 */
function RemoveComponent({ obtainData, hideAdd, href, name }:
  RemoveProps): JSX.Element {

  async function removeRequest() {

    const response = await fetch(href, {
      method: 'DELETE',
      headers: new Headers({
        'X-XSRF-TOKEN': csrfToken()
      })
    });

    if (response.ok) {

      alert(name + ' was removed.');
      obtainData();

    } else if (response.status === HttpStatus.UNAUTHORIZED) {
      alert('Incorrect user name or password');
    } else if (response.status === HttpStatus.FORBIDDEN) {
      alert('You do not have permission to delete ' + name + '.');
    } else {
      console.log(response.status)
      problemConnecting();
    }

  }

  function handleClick() {
    hideAdd();
    if (window.confirm("Are you sure you would like to remove " + name +
      "? This will also delete all of the harvested data assoicated with " + name + ".")) {
      removeRequest();
    }
  }


  return (

    <button data-testid='rm-button' onClick={handleClick}>{'Remove ' + name}</button>

  );

}

RemoveComponent.propTypes = {

  hideAdd: PropTypes.func.isRequired,

  obtainData: PropTypes.func.isRequired,

  name: PropTypes.string.isRequired,

  href: PropTypes.string.isRequired,
}


export default RemoveComponent;
