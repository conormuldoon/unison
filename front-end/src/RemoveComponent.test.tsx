import "@testing-library/jest-dom/extend-expect";
import fetchMock from 'fetch-mock';
import React from 'react';
import ReactDOM from 'react-dom';
import { fireEvent, render } from "@testing-library/react";
import RemoveComponent from './RemoveComponent';
import HttpStatus from 'http-status-codes';

const location = { name: 'UCD', links: { self: '/location/UCD' } };

const linksProperty = {
  "name": "UCD",
  "_links": {

    "self": {
      "href": "http://localhost:8080/locationCollection/UCD"
    }
  }

};

it('renders without crashing', async () => {

  const div = document.createElement('div');

  ReactDOM.render(<RemoveComponent href={linksProperty._links['self'].href} name={linksProperty.name}
    obtainData={() => { }} hideDisplay={() => { }}  />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('mathes snapshot', () => {


  const { container } = render(<RemoveComponent name={linksProperty.name} href={linksProperty._links['self'].href} obtainData={() => { }} hideDisplay={() => { }}  />);

  expect(container).toMatchSnapshot();

});


it('handles remove location', (done) => {

  fetchMock.delete('end:/' + location.name, HttpStatus.OK);


  const alertSpy = jest.spyOn(window, 'alert');
  alertSpy.mockImplementation(() => { });

  const confirmSpy = jest.spyOn(window, 'confirm');
  confirmSpy.mockImplementation(() => true);
  const hideDisplay = jest.fn();

  const obtainData = () => {
    expect(hideDisplay).toHaveBeenCalledTimes(1);
    expect(alert).toBeCalledWith(location.name + ' was removed.');
    done();
  };

  const component = <RemoveComponent name={linksProperty.name} href={linksProperty._links['self'].href} obtainData={obtainData} hideDisplay={hideDisplay}  />;
  const { getByText } = render(component);

  fireEvent.click(getByText('Remove ' + location.name));

  fetchMock.restore();
  confirmSpy.mockClear();
  alertSpy.mockClear();

});


