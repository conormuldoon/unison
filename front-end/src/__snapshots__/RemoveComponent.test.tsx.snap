// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathes snapshot 1`] = `
<div>
  <button
    data-testid="rm-button"
  >
    Remove UCD
  </button>
</div>
`;
